### ISSUE TITLE
Problems running under Windows 2008 R2 (0.6.0)

### ISSUE THREAD

**Issue by rabidpitbull** [rgl/packer-provisioner-windows-update#24]
I had an issue almost immediately, with the latest version of the provisioner on Windows 2008R2 (where I still have a need to generate test images and templates, due to a rather large number of servers in our environment that have not been upgraded...)

In the windows-update.ps1 file - I was getting an error where you were setting the updateDownloader priority to 4.  It appears that even though the Microsoft documentation implies that there is a dpExtraHigh settings, that it was not implemented in all their OS.

I was able to get past it by wrapping the statement in a try-catch.
```powershell
    try {
        $updateDownloader.Priority = 4 # 1 (dpLow), 2 (dpNormal), 3 (dpHigh), 4 (dpExtraHigh).
    } catch {}
```

Additionally, in both of the Powershell files in update folder, the use of ScriptStackTrace, within your trap blocks, causes issues on servers where Powershell 2.0 is still installed.  I was able to generate a meaningful error by utilizing properties of $error.InvocationInfo to capture the pertinent areas there were issues.
```powershell
    $err = $_
    Write-Output "ERROR: $($err)"
    if ($PSVersionTable.PSVersion.Major -gt 2) {
        Write-Output (($err.ScriptStackTrace -split '\r?\n') -replace '^(.*)$','ERROR: $1')
    } else {
        Write-Output "ERROR IN     : $($err.InvocationInfo.ScriptName)"
        Write-Output "LINE NUMBER  : $($err.InvocationInfo.ScriptLineNumber)"
        Write-Output "OFFSET       : $($err.InvocationInfo.OffsetInLine)"
        Write-Output "COMMAND      : $($err.InvocationInfo.MyCommand)"
        Write-Output "LINE         : $($err.InvocationInfo.Line)"
    }
    Write-Output (($err.Exception.ToString() -split '\r?\n') -replace '^(.*)$','ERROR EXCEPTION: $1')
```

Finally, the last issue I encountered was more cosmetic.  On these servers, I was running into an issue where Powershell was breaking up output longer than 79 characters into separate lines of output. Such as:

    virtualbox-iso: ERROR: Exception setting "Priority": "The parameter is incorrect. (Exception fr
    virtualbox-iso: om HRESULT: 0x80070057 (E_INVALIDARG))"

Instead of

    virtualbox-iso: ERROR: Exception setting "Priority": "The parameter is incorrect. (Exception from HRESULT: 0x80070057 (E_INVALIDARG))"

  I was able to remedy that by adding the following at the 'top' of the .ps1 files.

```powershell
$Host.UI.RawUI.BufferSize = New-Object Management.Automation.Host.Size (500, 25)
```

By and large, my testing with 0.6.0 has been very positive - hacking I had to do on the provisioner.go file to achieve stability on older OS versions is no longer needed.

The only other piece of info that I would offer up, is that I do have a stable, and working version of 
elevated-template.ps1 that configures the scheduled task directly, rather than importing a pre-baked .XML file.  This was developed based on some strangeness (the root of which I can't seem to remember...) where the .XML import would fail, when everything seemed to be working as it should.

I am extremely excited to use your provisioner to build images/templates.  It is a definite improvement on the sort of blind process of having the windows update logic built into the Autounattend.xml file.

If any or all of these are of interest, I can certainly cobble together a pull request.

Keep up the great work.
Cheers.

**Comment by tvories** [comment]
I am having I believe the same issue with Server 2008 R2.  I am using the VMware ISO provisioner with packer 1.2.5 and packer-provisioner-windows-update v0.6.0

``` log
2018/10/16 10:29:04 ui:     vmware-vmx: Found Windows update (2018-10-09; 8.93 MB): 2018-10 Update for Windows Server 2008 R2 for x64-based Systems (KB3177467)
2018/10/16 10:29:04 ui:     vmware-vmx: Downloading Windows updates (160 updates; 1418.86 MB)...
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR: Value does not fall within the expected range.
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR: at <ScriptBlock>, C:\Windows\Temp\packer-windows-update.ps1: line 195
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR: at <ScriptBlock>, <No file>: line 1
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR EXCEPTION: System.ArgumentException: Value does not fall within the expected range.
2018/10/16 10:29:04 packer-provisioner-windows-update.exe: 2018/10/16 10:29:04 [INFO] 20711 bytes written for 'stdout'
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR EXCEPTION:    at System.Management.Automation.ComInterop.ComRuntimeHelpers.CheckThrowException(Int32 hresult, ExcepInfo& excepInfo, ComMethodDesc 
method, Object[] args, UInt32 argErr)
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR EXCEPTION:    at CallSite.Target(Closure , CallSite , ComObject
 , Int32 )
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR EXCEPTION:    at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR EXCEPTION:    at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR EXCEPTION:    at System.Management.Automation.Interpreter.DynamicInstruction`3.Run(InterpretedFrame frame)
2018/10/16 10:29:04 ui:     vmware-vmx: ERROR EXCEPTION:    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)

2018/10/16 10:29:04 packer-provisioner-windows-update.exe: 2018/10/16 10:29:04 [INFO] RPC client: Communicator ended with: 1

2018/10/16 10:29:04 [INFO] (telemetry) ending windows-update
```

**Comment by rgl** [comment]
Thanks for the feedback! I didn't known about BufferSize, I will definitely use it!

Only having PowerShell 2 is a real bummer, and I don't want to add support for it, essentially, because 2008R2 and 2012R2 are being/near mainstream EOL. Instead, when I'm creating a base 2012R2 box at [windows-2016-vagrant provision-powershell.ps1](https://github.com/rgl/windows-2016-vagrant/blob/wip/provision-powershell.ps1) I'm installing PowerShell 5.1 before doing anything else.

So please install PowerShell 5.1 before using the provisioner.

Not having $updateDownloader.Priority is a real bummer too, I will wrap it in a try catch. But bear in mind that maintaining a script for old versions of windows is not really in my plans and is a really bummer for me, and I will probably not do more than this Priority change. 

I will also create the scheduled task using proper cmdlets (like I'm doing at https://github.com/rgl/windows-2016-vagrant/blob/wip/libvirt-fix-cpu-driver.ps1#L104).

**Comment by rabidpitbull** [comment]
I appreciate your position, I’ll probably just have to fork the project for my own needs then.

I find that I have to do a lot of defensive programming as I build test images in support of an environment with a lot of older installs (2008r2, 2012, 2012r2).  It’s not a question of only having PS 2, it’s a question of making sure Chef scripts work on machines that still have PS 2...  we’re moving to get folks to migrate, but in a larger enterprise some folks don’t want to fly too close to the sun...

Thanks.
—
Tom

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by tvories** [comment]
My results were using PowerShell 5.1 as a heads up.  I don't think you should worry too much about PowerShell 2.0, but it would be nice to fix the $updateDownloader.Priority.  I'd be happy to submit pull requests on that end.

**Comment by tvories** [comment]
This appears to be fixed with [0.6.1](https://github.com/rgl/packer-provisioner-windows-update/releases/tag/v0.6.1)!
